main/src/ca/uwaterloo/flix/language/ast/TypedAst.scala

```diff
@@ -122,6 +122,8 @@ object TypedAst {
 
     case class Apply(exp: Expr, exps: List[Expr], tpe: Type, eff: Type, loc: SourceLocation) extends Expr
 
+    case class ApplyDef(symUse: Ast.DefSymUse, exps: List[Expr], ftpe: Type, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
```

Comment:
Because we have lost the type of the def itself which was located in `Expr.Def`. I think @JonathanStarup can explain a little better

Reply:
I am 95% sure this type can be reconstructed from the available information, e.g. from the types of exps, from the eff and tpe.

Reply:
I think there was an issue with this example:

```
def f(_: Int32, _: Int32): Bool \ ef = ???

f(1, g()) // g returns Int but does IO
```

The `f` has the variable `ef` and the call has `IO` effect so the call to `f` has type `ef + IO` then we cannot reconstruct the type of `f` because `ef` unifies with `IO` but it could also just be `ef`.

We can, however, easily get rid of the result type that we carry everywhere

Reply:
We had something like this before and it was an absolute mess because the types would come out of sync.

My question is: What information is it that you need? The type scheme of f? The instantiated type of f? Something else?

Strictly speaking, in your example, the effect of `f`can be found by simply looking up its type scheme.

Reply:
I think I understand the problem. I will give you a chance to think carefully about it, and describe it, so you understand it too :)

Reply:
So i shouldn't explain? 

> I am 95% sure this type can be reconstructed from the available information, e.g. from the types of exps, from the eff and tpe.

But yea this is not true in monomorph. we don't know the effect of the def to specialize to without more information (ftpe or intrinsicEff or something third)